fix: replace `IMarkdownSourceCode` with `MarkdownSourceCode`

lumirlumir opened the pull request:
Hello,

This PR addresses issue https://github.com/eslint/rewrite/issues/175.

I also added `rollup-plugin-delete` to fix the build issues in CI. It's the same plugin used in [JSON](https://github.com/eslint/json/blob/main/rollup.config.js)'s `rollup.config.js`.

## Summary

When using `IMarkdownSourceCode`, we don't have access to methods like `getParent`, `getAncestors`, `getText`, and `lines`. To fix this, I replaced it with `MarkdownSourceCode`, which does provide those methods. This approach is already being used in the [JSON plugin](https://github.com/eslint/json/blob/main/src/types.ts#L77-L92).

## Fixes

Fixes: https://github.com/eslint/rewrite/issues/175.

nzakas reviewed (dismissed):
LGTM. Thanks!

Would like @fasttime to review before merging.

fasttime commented on src/types.ts:
It looks like the `IMarkdownSourceCode` type is no longer required. I think we can remove it and replace it with `MarkdownSourceCode` in the tests.

lumirlumir replied:
I've resolved merge conflict!

---

`IMarkdownSourceCode` is still being used in the `MarkdownSourceCode` class, like in the following snippet:

https://github.com/eslint/markdown/blob/main/src/language/markdown-source-code.js#L137-L141

```js
/**
 * Markdown Source Code Object
 * @implements {IMarkdownSourceCode}
 */
export class MarkdownSourceCode extends TextSourceCodeBase {
```

Would it be okay to replace it with an inline type like `{TextSourceCode<{LangOptions: JSONLanguageOptions, RootNode: DocumentNode, SyntaxElementWithLoc: JSONSyntaxElement, ConfigNode: Token}>}` as we did in [JSON](https://github.com/eslint/json/blob/main/src/languages/json-source-code.js#L66-L70)?

fasttime replied:
Yes, that's the best solution I think.

lumirlumir replied:
Sounds good. I'll go ahead and update the tests and remove `IMarkdownSource`.

fasttime reviewed (commented):
There's a merge conflict now.

lumirlumir commented:
@fasttime 

I've removed the `IMarkdownSourceCode` interface and replaced it with `MarkdownSourceCode` in https://github.com/eslint/markdown/pull/336/commits/6330ad423163ed3f2055ba6f8271d978e6507f80!

fasttime commented on src/language/markdown-source-code.js:
```diff
@@ -26,13 +26,13 @@ import { findOffsets } from "../util.js";
 /** @typedef {import("@eslint/core").TraversalStep} TraversalStep */
 /** @typedef {import("@eslint/core").VisitTraversalStep} VisitTraversalStep */
-/** @typedef {import("@eslint/core").TextSourceCode} TextSourceCode */
+/** @typedef {import("@eslint/core").TextSourceCode<{LangOptions: MarkdownLanguageOptions, RootNode: RootNode, SyntaxElementWithLoc: MarkdownNode, ConfigNode: { value: string; position: SourceLocation }}>} TextSourceCode */
```
```suggestion
/** @import { TextSourceCode } from "@eslint/core"; */
```

Maybe we could use an `@import` tag here to import `TextSourceCode` as a generic type and add the type parameters on the annotation of `MarkdownSourceCode` instead? Like it's done in `@eslint/json`:

https://github.com/eslint/json/blob/dce4601b1a40ceaeb4b61fbcbef0170a67b73e37/src/languages/json-source-code.js#L68

lumirlumir replied:
[JSON](https://github.com/eslint/json/blob/dce4601b1a40ceaeb4b61fbcbef0170a67b73e37/src/languages/json-source-code.js#L22-L27) only uses `@import` and doesn’t use `@typedef`.

```js
/**
 * @import { DocumentNode, Node, Token } from "@humanwhocodes/momoa";
 * @import { SourceLocation, FileProblem, DirectiveType, RulesConfig, TextSourceCode } from "@eslint/core";
 * @import { JSONSyntaxElement } from "../types.ts";
 * @import { JSONLanguageOptions } from "./json-language.js";
 */
```

So, would it be better to convert all `@typedef` to `@import`?

Or should we leave the other `@typedef` statements as they are and only use `@import` for `TextSourceCode`?

fasttime replied:
> So, would it be better to convert all `@typedef` to `@import`?
> 
> Or should we leave the other `@typedef` statements as they are and only use `@import` for `TextSourceCode`?

`@eslint/json` uses additional logic to handle `@import` tags during the build in [tools/dedupe-types.js](https://github.com/eslint/json/blob/main/tools/dedupe-types.js). If we want switch to `@import` everywhere, we should also update the build scripts in this repo, which is probably a good idea but out of scope for this change.

Alternatively, we could keep using the `@typedef` syntax with `@template` to re-export `TextSourceCode` as a generic type, like with `MarkdownRuleDefinition` here:

https://github.com/eslint/markdown/blob/13dfd5ec279d1db3cbd1ef2b6bf5d00799371090/src/index.js#L22-L25

lumirlumir replied:
I'd like to use `@typedef` and `@template` for this PR.

Once all my current PRs are resolved, I'll open another one to switch to the `@import` syntax!

fasttime reviewed (commented):
Note that there is a new merge conflict now in `src/language/markdown-source-code.js`.

lumirlumir commented:
@fasttime 

I've exported the `TextSourceCode` type in https://github.com/eslint/markdown/pull/336/commits/6a90d10e9bd6f30ec7946791dc8734bcf734719d.  

It's now generic, with default options included!

fasttime commented on src/language/markdown-source-code.js:
```diff
@@ -26,14 +26,18 @@ import { findOffsets } from "../util.js";
-/** @typedef {import("../types.ts").IMarkdownSourceCode} IMarkdownSourceCode */
+/**
+ * @typedef {import("@eslint/core").TextSourceCode<Options>} TextSourceCode<Options>
+ * @template {MarkdownTextSourceCodeOptions} [Options=MarkdownTextSourceCodeOptions]
```
```suggestion
 * @template {SourceCodeBaseTypeOptions} [Options=SourceCodeBaseTypeOptions]
```

It's better to maintain the default type parameter as defined in `@eslint/core` for consistency with `@eslint/js` and also to avoid confusion.

Now, `SourceCodeBaseTypeOptions` isn't exported by `@eslint/core` (https://github.com/eslint/rewrite/blob/core-v0.13.0/packages/core/src/types.ts#L854-L871), but it can be defined as needed:

```ts
interface SourceCodeBaseTypeOptions {
    LangOptions: LanguageOptions;
    RootNode: unknown;
    SyntaxElementWithLoc: unknown;
    ConfigNode: unknown;
};
```

or deduced from `TextSourceCode`:

```ts
type SourceCodeBaseTypeOptions = TextSourceCode extends TextSourceCode<infer Options> ? Options : never;
```

lumirlumir replied:
I'll take the former suggestion.  

After I implement it, I'd like to make a PR to `@eslint/core` to export `SourceCodeBaseTypeOptions`.  

If the suggestion is accepted, I'll refactor it to use `SourceCodeBaseTypeOptions` from `@eslint/core`.

fasttime replied:
> After I implement it, I'd like to make a PR to `@eslint/core` to export `SourceCodeBaseTypeOption`

Feel free :)

fasttime commented on src/language/markdown-source-code.js:
```diff
@@ -137,7 +141,7 @@ function extractInlineConfigCommentsFromHTML(node) {
 /**
  * Markdown Source Code Object
- * @implements {IMarkdownSourceCode}
+ * @implements {TextSourceCode}
```
```suggestion
 * @implements {TextSourceCode<{ LangOptions: MarkdownLanguageOptions; RootNode: RootNode; SyntaxElementWithLoc: Node; ConfigNode: { value: string; position: SourceLocation }; }>}
```

Could we specify the type param options here like it's done in `@eslint/json`: https://github.com/eslint/json/blob/dce4601b1a40ceaeb4b61fbcbef0170a67b73e37/src/languages/json-source-code.js#L68

lumirlumir replied:
Oops, I missed that. thank you.

fasttime reviewed (approved):
LGTM, thanks!
